Zend/zend_inheritance.c

```diff
@@ -459,6 +512,56 @@ static inheritance_status zend_perform_covariant_class_type_check(
 	return have_unresolved ? INHERITANCE_UNRESOLVED : INHERITANCE_ERROR;
 }
 
+/* checks that the child type (being unique) is a subtype of each member of the parent intersection */
+static inheritance_status zend_is_single_type_subtype_intersection(
```

Comment:
I'm still super confused by the variance code. The way this is currently structured for union types is that the problem is split into two parts: If we have a check on `X subtype Y` then we first check that `X_1|...|X_n` is subtype `Y` by checking that each `X_i` is subtype `Y`. That second part is then another check for `X_i` subtype `Y_1|...|Y_n`, which is true if `X_i` is subtype of some `Y_j`.

I would expect that for intersection types the code would be factored in exactly the same way, reusing the existing functions, just flipping the `&&` logic for `||` and vice versa. For example, `zend_perform_covariant_class_type_check` is the function responsible for testing one simple LHS class type against a complex RHS type. I would expect intersection handling to be added to that function by adjusting part at the end that walks the type list to short-circuit in the inverse case, etc.

Reply:
So I initially tried to reuse the current functions, but was hitting too many edge cases and when I tried to fix them I impacted the union variance, so I decided to split it up completely, I'll try and see if I can unify them again.